File: lib/javascript/src/subscription.ts
import {
	assertChecksum,
	assertLength,
	BLOCK_SIZE,
	isZero,
	withChecksum,
} from "./bytes";
import { decodeDate, encodeDate, type Date16Bit } from "./date";
import { decodeTime, encodeTime, type Time } from "./time";

export type Subscription = {
	startsAt: Date16Bit;
	endsAt: Date16Bit;
	unknown1: Uint8Array;
	unknown2: Uint8Array;
	lastUsedAt?: Date16Bit & Time;
};

export function decodeSubscription(block: Uint8Array): Subscription {
	assertLength(block, BLOCK_SIZE, "subscription block");
	assertChecksum(block);
	const lastUsed = block.subarray(10, 15);
	return {
		startsAt: decodeDate(block.subarray(0, 2)),
		endsAt: decodeDate(block.subarray(2, 4)),
		unknown1: block.slice(4, 6),
		unknown2: block.slice(6, 10),
		lastUsedAt: isZero(lastUsed)
			? undefined
			: {
					...decodeDate(lastUsed.subarray(0, 2)),
					...decodeTime(lastUsed.subarray(2, 5)),
				},
	};
}

export function encodeSubscription(subscription: Subscription): Uint8Array {
	assertLength(subscription.unknown1, 2, "unknown1");
	assertLength(subscription.unknown2, 4, "unknown2");
	const block = new Uint8Array(BLOCK_SIZE);
	block.set(encodeDate(subscription.startsAt), 0);
	block.set(encodeDate(subscription.endsAt), 2);
	block.set(subscription.unknown1, 4);
	block.set(subscription.unknown2, 6);
	if (subscription.lastUsedAt) {
		block.set(encodeDate(subscription.lastUsedAt), 10);
		block.set(encodeTime(subscription.lastUsedAt), 12);
	}
	return withChecksum(block);
}

File: lib/rust/src/subscription.rs
use crate::bytes::{BLOCK_SIZE, Block, check_checksum, chunk, is_zero, with_checksum};
use crate::date::Date;
use crate::date_time::DateTime;
use crate::error::Result;

/// Block 13 or 17 of a personal card, copied in 14 and 18.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Subscription {
	/// First day of validity.
	pub starts_at: Date,
	/// Last day of validity.
	pub ends_at: Date,
	/// Bytes 4 and 5, always `0000` so far.
	pub unknown1: [u8; 2],
	/// Bytes 6 to 9.
	pub unknown2: [u8; 4],
	/// Last use the pass recorded; absent while unused.
	pub last_used_at: Option<DateTime>,
}

impl Subscription {
	/// Decodes block 13 or 17.
	///
	/// # Errors
	/// [`Error::Checksum`](crate::Error::Checksum) or the errors of [`Date::decode`] and [`DateTime::decode`].
	pub fn decode(block: &Block) -> Result<Self> {
		check_checksum(block)?;
		let last_used = &block[10..15];
		Ok(Self {
			starts_at: Date::decode(chunk(block, 0))?,
			ends_at: Date::decode(chunk(block, 2))?,
			unknown1: chunk(block, 4),
			unknown2: chunk(block, 6),
			last_used_at: if is_zero(last_used) {
				None
			} else {
				Some(DateTime::decode(chunk(block, 10))?)
			},
		})
	}

	/// Encodes into block 13 or 17 with its checksum.
	///
	/// # Errors
	/// The errors of [`Date::encode`] and [`DateTime::encode`].
	pub fn encode(self) -> Result<Block> {
		let mut block = [0; BLOCK_SIZE];
		block[..2].copy_from_slice(&self.starts_at.encode()?);
		block[2..4].copy_from_slice(&self.ends_at.encode()?);
		block[4..6].copy_from_slice(&self.unknown1);
		block[6..10].copy_from_slice(&self.unknown2);
		if let Some(last_used_at) = self.last_used_at {
			block[10..15].copy_from_slice(&last_used_at.encode()?);
		}
		Ok(with_checksum(block))
	}
}
